File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/conformance/events_runner.rb
# frozen_string_literal: true

require "json"
require_relative "result"
require_relative "../audit/memory_sink"
require_relative "../core/pressure_config"
require_relative "../core/pressure_engine"
require_relative "../core/clock"

module IAIso
  module Conformance
    module EventsRunner
      TOLERANCE = 1e-9

      module_function

      def run(spec_root)
        path = File.join(spec_root, "events", "vectors.json")
        doc = ::JSON.parse(File.read(path))
        doc["vectors"].map { |v| run_one(v) }
      end

      def run_one(v)
        name = v["name"] || "?"
        kw = {}
        if v["config"].is_a?(Hash)
          c = v["config"]
          %w[escalation_threshold release_threshold dissipation_per_step
             dissipation_per_second token_coefficient tool_coefficient
             depth_coefficient].each do |f|
            kw[f.to_sym] = c[f].to_f if c.key?(f) && c[f].is_a?(Numeric)
          end
          kw[:post_release_lock] = c["post_release_lock"] if [true, false].include?(c["post_release_lock"])
        end
        cfg = IAIso::Core::PressureConfig.new(**kw)
        clock_seq = (v["clock"].is_a?(Array) ? v["clock"] : [0.0]).map(&:to_f)
        clk = IAIso::Core::ScriptedClock.new(clock_seq)

        sink = IAIso::Audit::MemorySink.new
        exec_id = v["execution_id"] || "?"
        engine =
          begin
            IAIso::Core::PressureEngine.new(
              config: cfg,
              options: IAIso::Core::EngineOptions.new(
                execution_id: exec_id, audit_sink: sink, clock: clk, timestamp_clock: clk,
              ),
            )
          rescue => e
            return VectorResult.fail("events", name, "engine init failed: #{e.message}")
          end

        reset_after_step = v["reset_after_step"]
        steps = v["steps"] || []
        steps.each_with_index do |step, i|
          if step["reset"]
            engine.reset
          else
            kw2 = {}
            kw2[:tokens] = step["tokens"].to_i if step.key?("tokens")
            kw2[:tool_calls] = step["tool_calls"].to_i if step.key?("tool_calls")
            kw2[:depth] = step["depth"].to_i if step.key?("depth")
            kw2[:tag] = step["tag"] if step.key?("tag")
            engine.step(IAIso::Core::StepInput.build(**kw2))
          end
          # 1-based: reset_after_step = N triggers after running step N
          if reset_after_step && (i + 1) == reset_after_step
            engine.reset
          end
        end

        got = sink.events
        expected = v["expected_events"] || []
        if got.size != expected.size
          return VectorResult.fail("events", name, "event count: got #{got.size}, want #{expected.size}")
        end
        expected.each_with_index do |exp, i|
          actual = got[i]
          if exp["schema_version"] && exp["schema_version"] != "" && exp["schema_version"] != actual.schema_version
            return VectorResult.fail("events", name, "event #{i} schema_version: got #{actual.schema_version}, want #{exp["schema_version"]}")
          end
          if exp["execution_id"] && exp["execution_id"] != "" && exp["execution_id"] != actual.execution_id
            return VectorResult.fail("events", name, "event #{i} execution_id: got #{actual.execution_id}, want #{exp["execution_id"]}")
          end
          if exp["kind"] && exp["kind"] != actual.kind
            return VectorResult.fail("events", name, "event #{i} kind: got #{actual.kind}, want #{exp["kind"]}")
          end
          if exp["data"].is_a?(Hash)
            unless data_matches?(actual.data, exp["data"])
              return VectorResult.fail("events", name, "event #{i} data mismatch: got #{actual.data}, want #{exp["data"]}")
            end
          end
        end
        VectorResult.pass("events", name)
      end

      def data_matches?(actual, want)
        # Actual data uses string keys; compare each key in want.
        want.each do |k, w|
          got = actual[k] || actual[k.to_sym]
          return false unless value_equal?(got, w)
        end
        true
      end

      def value_equal?(actual, want)
        return actual.nil? if want.nil?
        return false if actual.nil?
        return actual == want if [true, false].include?(want)
        if want.is_a?(Numeric)
          return false unless actual.is_a?(Numeric)
          return (actual.to_f - want.to_f).abs <= TOLERANCE
        end
        return actual == want if want.is_a?(String)
        if want.is_a?(Array)
          return false unless actual.is_a?(Array) && actual.size == want.size
          want.each_with_index do |w, i|
            return false unless value_equal?(actual[i], w)
          end
          return true
        end
        if want.is_a?(Hash)
          return false unless actual.is_a?(Hash)
          want.each do |k, w|
            a = actual[k] || actual[k.to_sym]
            return false unless value_equal?(a, w)
          end
          return true
        end
        false
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Conformance/EventsRunner.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json.Nodes;
using Iaiso.Audit;
using Iaiso.Core;

namespace Iaiso.Conformance;

internal static class EventsRunner
{
    public static List<VectorResult> Run(string specRoot)
    {
        var path = Path.Combine(specRoot, "events", "vectors.json");
        var doc = JsonNode.Parse(File.ReadAllBytes(path))!.AsObject();
        var vectors = doc["vectors"]!.AsArray();
        var results = new List<VectorResult>(vectors.Count);
        foreach (var v in vectors) results.Add(RunOne(v!.AsObject()));
        return results;
    }

    private static VectorResult RunOne(JsonObject v)
    {
        string name = v["name"]!.GetValue<string>();

        var cb = PressureConfig.CreateBuilder();
        if (v["config"] is JsonObject c)
        {
            ApplyDouble(c, "escalation_threshold", x => cb.EscalationThreshold = x);
            ApplyDouble(c, "release_threshold", x => cb.ReleaseThreshold = x);
            ApplyDouble(c, "dissipation_per_step", x => cb.DissipationPerStep = x);
            ApplyDouble(c, "dissipation_per_second", x => cb.DissipationPerSecond = x);
            ApplyDouble(c, "token_coefficient", x => cb.TokenCoefficient = x);
            ApplyDouble(c, "tool_coefficient", x => cb.ToolCoefficient = x);
            ApplyDouble(c, "depth_coefficient", x => cb.DepthCoefficient = x);
            if (c["post_release_lock"] is JsonValue prl && prl.TryGetValue<bool>(out var b))
                cb.PostReleaseLock = b;
        }
        var cfg = cb.Build();

        double[] clockSeq;
        if (v["clock"] is JsonArray clkArr)
        {
            clockSeq = new double[clkArr.Count];
            for (int i = 0; i < clkArr.Count; i++) clockSeq[i] = clkArr[i]!.GetValue<double>();
        }
        else
        {
            clockSeq = new[] { 0.0 };
        }
        var clock = new ScriptedClock(clockSeq);

        var sink = new MemorySink();
        string execId = v["execution_id"]!.GetValue<string>();
        PressureEngine engine;
        try
        {
            engine = new PressureEngine(cfg, new EngineOptions
            {
                ExecutionId = execId,
                AuditSink = sink,
                Clock = clock,
                TimestampClock = clock,
            });
        }
        catch (Exception e)
        {
            return VectorResult.Fail("events", name, "engine init failed: " + e.Message);
        }

        int? resetAfterStep = v["reset_after_step"] is JsonValue rs
            && rs.TryGetValue<int>(out var rsi) ? rsi : null;

        var steps = v["steps"] as JsonArray ?? new JsonArray();
        for (int i = 0; i < steps.Count; i++)
        {
            var step = steps[i]!.AsObject();
            if (step["reset"] is JsonValue rv && rv.TryGetValue<bool>(out var resetB) && resetB)
            {
                engine.Reset();
            }
            else
            {
                var si = new StepInput();
                if (step["tokens"] is JsonValue tv) si.Tokens = tv.GetValue<long>();
                if (step["tool_calls"] is JsonValue tcv) si.ToolCalls = tcv.GetValue<long>();
                if (step["depth"] is JsonValue dv) si.Depth = dv.GetValue<long>();
                if (step["tag"] is JsonValue tg && tg.TryGetValue<string>(out var tag)) si.Tag = tag;
                engine.Step(si);
            }
            if (resetAfterStep.HasValue && (i + 1) == resetAfterStep.Value)
            {
                engine.Reset();
            }
        }

        var got = sink.Events();
        var expected = v["expected_events"]!.AsArray();
        if (got.Count != expected.Count)
            return VectorResult.Fail("events", name,
                $"event count: got {got.Count}, want {expected.Count}");

        for (int i = 0; i < expected.Count; i++)
        {
            var exp = expected[i]!.AsObject();
            var actual = got[i];
            if (exp["schema_version"] is JsonValue sv && sv.TryGetValue<string>(out var svs)
                && !string.IsNullOrEmpty(svs) && svs != actual.SchemaVersion)
                return VectorResult.Fail("events", name,
                    $"event {i} schema_version: got {actual.SchemaVersion}, want {svs}");
            if (exp["execution_id"] is JsonValue ev && ev.TryGetValue<string>(out var evs)
                && !string.IsNullOrEmpty(evs) && evs != actual.ExecutionId)
                return VectorResult.Fail("events", name,
                    $"event {i} execution_id: got {actual.ExecutionId}, want {evs}");
            string wantKind = exp["kind"]!.GetValue<string>();
            if (wantKind != actual.Kind)
                return VectorResult.Fail("events", name,
                    $"event {i} kind: got {actual.Kind}, want {wantKind}");
            if (exp["data"] is JsonObject expData)
            {
                if (!DataMatches(actual.Data, expData))
                    return VectorResult.Fail("events", name,
                        $"event {i} data mismatch: got {ActualToJson(actual.Data)}, want {expData.ToJsonString()}");
            }
        }
        return VectorResult.Pass("events", name);
    }

    private static string ActualToJson(IReadOnlyDictionary<string, object?> data)
    {
        // For diagnostic output only
        var obj = new JsonObject();
        foreach (var kv in data)
        {
            obj[kv.Key] = kv.Value switch
            {
                null => null,
                string s => JsonValue.Create(s),
                bool b => JsonValue.Create(b),
                long l => JsonValue.Create(l),
                int i => JsonValue.Create(i),
                double d => JsonValue.Create(d),
                _ => JsonValue.Create(kv.Value.ToString()),
            };
        }
        return obj.ToJsonString();
    }

    private static bool DataMatches(IReadOnlyDictionary<string, object?> actual, JsonObject want)
    {
        foreach (var entry in want)
        {
            object? got = actual.TryGetValue(entry.Key, out var v) ? v : null;
            if (!ValueEqual(got, entry.Value)) return false;
        }
        return true;
    }

    private static bool ValueEqual(object? actual, JsonNode? want)
    {
        if (want is null) return actual is null;
        if (actual is null) return false;
        if (want is JsonValue v)
        {
            if (v.TryGetValue<bool>(out var b))
                return actual is bool ab && ab == b;
            if (v.TryGetValue<double>(out var d))
            {
                if (actual is double ad) return Math.Abs(ad - d) <= ConformanceRunner.Tolerance;
                if (actual is long al) return Math.Abs((double)al - d) <= ConformanceRunner.Tolerance;
                if (actual is int ai) return Math.Abs((double)ai - d) <= ConformanceRunner.Tolerance;
                return false;
            }
            if (v.TryGetValue<string>(out var s))
                return actual is string @as && @as == s;
        }
        if (want is JsonArray arr)
        {
            if (actual is not System.Collections.IEnumerable e) return false;
            var aList = new List<object?>();
            foreach (var item in e) aList.Add(item);
            if (aList.Count != arr.Count) return false;
            for (int i = 0; i < arr.Count; i++)
                if (!ValueEqual(aList[i], arr[i])) return false;
            return true;
        }
        if (want is JsonObject obj)
        {
            if (actual is not IReadOnlyDictionary<string, object?> aDict) return false;
            foreach (var kv in obj)
            {
                object? got = aDict.TryGetValue(kv.Key, out var x) ? x : null;
                if (!ValueEqual(got, kv.Value)) return false;
            }
            return true;
        }
        return false;
    }

    private static void ApplyDouble(JsonObject obj, string key, Action<double> setter)
    {
        if (obj[key] is JsonValue v && v.TryGetValue<double>(out var d)) setter(d);
    }
}
